pandas/core/dtypes/dtypes.py

```diff
@@ -665,17 +666,16 @@ def __new__(cls, subtype=None):
             except TypeError:
                 raise ValueError("could not construct IntervalDtype")
 
-        if subtype is None:
-            u = object.__new__(cls)
-            u.subtype = None
-            return u
-
         try:
-            return cls._cache[str(subtype)]
+            # GH 18980: need to combine since str and hash individually may not
+            # be unique, e.g. str(CategoricalDtype) always returns 'category',
+            # and hash(np.dtype('<m8')) == hash(np.dtype('<m8[ns]'))
+            key = ''.join([str(subtype), str(hash(subtype))])
```

Comment:
what are you doing here?

Reply:
Trying to find a way create unique keys for each subtype that doesn't result in unwanted collisions.  The previous behavior was essentially `key=str(subtype)`.  That doesn't work for `CategoricalDtype` subtypes though, since `str(CategoricalDtype(*))` always returns `'category'`.  You then end up with a collision if you try to create a second `IntervalDtype` with `CategoricalDtype` subtype, and get back an `IntervalDtype`  with attributes from the first `CategoricalDtype`:
```python
In [12]: idt1 = IntervalDtype(CategoricalDtype(list('abc'), True))

In [13]: idt2 = IntervalDtype(CategoricalDtype(list('wxyz'), False))

In [14]: idt2.subtype
Out[14]: CategoricalDtype(categories=['a', 'b', 'c'], ordered=True)
```

My initial thought to fix this was to use `hash` instead of `str`, so essentially `key = hash(subtype)`.  That also doesn't quite work, because some numpy dtypes hash to the same value even though they're different:
```python
In [2]: np.dtype('<m8') == np.dtype('<m8[ns]')
Out[2]: False

In [3]: hash(np.dtype('<m8')) == hash(np.dtype('<m8[ns]'))
Out[3]: True
```

I originally implemented it with `key = hash(subtype)`, and the tests failed on travis and appveyor, so it is something that comes up.

It looked like the dtypes that failed for `str` and `hash` were mutually exclusive, so making the key a combination of `str` and `hash` seemed like it would work.  Certainly open to other solutions though, as it's not the cleanest method.  Or should we just revert to `str` and explicitly not support `CategoricalDtype` as a subtype?  Not sure how often that would actually come up.